pkg/libartifact/store/store.go

```diff
@@ -0,0 +1,409 @@
+	index := specV1.Index{
+		MediaType: specV1.MediaTypeImageIndex,
+		Versioned: specs.Versioned{SchemaVersion: 2},
+	}
+	rawData, err := json.Marshal(&index)
+	if err != nil {
+		return err
+	}
+
+	return os.WriteFile(as.indexPath(), rawData, 0o644)
+}
+
+func (as ArtifactStore) indexPath() string {
+	return filepath.Join(as.storePath, indexName)
+}
+
+// getArtifacts returns an ArtifactList based on the artifact's store.  The return error and
+// unused opts is meant for future growth like filters, etc so the API does not change.
+func (as ArtifactStore) getArtifacts(ctx context.Context, _ *libartTypes.GetArtifactOptions) (libartifact.ArtifactList, error) {
+	var (
+		al libartifact.ArtifactList
+	)
+
+	lrs, err := layout.List(as.storePath)
+	if err != nil {
+		return nil, err
+	}
+	for _, l := range lrs {
+		imgSrc, err := l.Reference.NewImageSource(ctx, as.SystemContext)
+		if err != nil {
+			return nil, err
+		}
+		manifests, err := getManifests(ctx, imgSrc, nil)
+		if err != nil {
+			return nil, err
+		}
+		artifact := libartifact.Artifact{
+			Manifests: manifests,
+		}
+		if val, ok := l.ManifestDescriptor.Annotations[specV1.AnnotationRefName]; ok {
+			artifact.SetName(val)
+		}
+
+		al = append(al, &artifact)
+	}
+	return al, nil
+}
+
+// getManifests takes an imgSrc and starting digest (nil means "top") and collects all the manifests "under"
+// it.  this func calls itself recursively with a new startingDigest assuming that we are dealing with
+// and index list
+func getManifests(ctx context.Context, imgSrc types.ImageSource, startingDigest *digest.Digest) ([]manifest.OCI1, error) {
+	var (
+		manifests []manifest.OCI1
+	)
+	b, manifestType, err := imgSrc.GetManifest(ctx, startingDigest)
+	if err != nil {
+		return nil, err
+	}
+
+	// this assumes that there are only single, and multi-images
+	if !manifest.MIMETypeIsMultiImage(manifestType) {
+		// these are the keepers
+		mani, err := manifest.OCI1FromManifest(b)
+		if err != nil {
+			return nil, err
+		}
+		manifests = append(manifests, *mani)
+		return manifests, nil
+	}
+	// We are dealing with an oci index list
+	maniList, err := manifest.OCI1IndexFromManifest(b)
+	if err != nil {
+		return nil, err
+	}
+	for _, m := range maniList.Manifests {
+		iterManifests, err := getManifests(ctx, imgSrc, &m.Digest)
+		if err != nil {
+			return nil, err
+		}
+		manifests = append(manifests, iterManifests...)
+	}
+	return manifests, nil
+}
+
+func checkForEmptyStanzaFile(path string) error {
+	if _, err := os.Stat(path); err == nil {
+		return nil
+	}
+	return os.WriteFile(path, emptyStanza, 0644)
+}
+
+func determineManifestType(path string) (string, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+	// DetectContentType looks at the first 512 bytes
+	b := make([]byte, 512)
+	// Because DetectContentType will return a default value
+	// we don't sweat the error
+	_, _ = f.Read(b)
+	return http.DetectContentType(b), nil
```

Comment:
This should probably be passing `b[:n]`, where `n` is the first value returned by the `f.Read()` call.